zap/src/test/java/org/zaproxy/zap/extension/pscan/PassiveScanDataUnitTest.java

```diff
@@ -0,0 +1,138 @@
+
+    private Session session;
+
+    @BeforeEach
+    public void setup() {
+        Constant.messages = mock(I18N.class);
+        session = mock(Session.class);
+        doReturn(session).when(model).getSession();
+    }
+
+    @Test
+    public void shouldHaveAllTechSetByDefault() {
+        // Given / When
+        HttpMessage msg = createMessage();
+        PassiveScanData psd = new PassiveScanData(msg);
+        // Then
+        assertThat(psd.getTechSet(), is(equalTo(TechSet.AllTech)));
+    }
+
+    @Test
+    public void shouldHaveEmptyUserListByDefault() {
+        // Given / When
+        HttpMessage msg = createMessage();
+        PassiveScanData psd = new PassiveScanData(msg);
+        // Then
+        assertThat(psd.getUsers(), is(equalTo(Collections.emptyList())));
+    }
+
+    @Test
+    public void shouldNotHaveContextByDefault() {
+        // Given / When
+        HttpMessage msg = createMessage();
+        PassiveScanData psd = new PassiveScanData(msg);
+        // Then
+        assertNull(psd.getContext());
+        assertThat(psd.hasContext(), is(equalTo(false)));
+    }
+
+    @Test
+    public void shouldHaveContextIfMessageIncludedInOne() {
+        // Given
+        HttpMessage msg = createMessage();
+        Context context = mock(Context.class);
+        given(session.getContextsForUrl(msg.getRequestHeader().getURI().toString()))
+                .willReturn(asList(context));
+        // When
+        PassiveScanData psd = new PassiveScanData(msg);
+        // Then
+        assertThat(psd.getContext(), is(equalTo(context)));
+        assertThat(psd.hasContext(), is(equalTo(true)));
+    }
+
+    @Test
+    public void shouldUseFirstContextIfMessageApplicableToMultiple() {
+        // Given
+        HttpMessage msg = createMessage();
+        Context matchCtxOne = mock(Context.class);
+        Context matchCtxTwo = mock(Context.class);
+        given(session.getContextsForUrl(msg.getRequestHeader().getURI().toString()))
+                .willReturn(asList(matchCtxOne, matchCtxTwo));
+        // When
+        PassiveScanData psd = new PassiveScanData(msg);
+        // Then
+        assertThat(psd.getContext(), is(equalTo(matchCtxOne)));
+        assertThat(psd.hasContext(), is(equalTo(true)));
+    }
+
+    @Test
+    public void shouldUseTechSetOfFirstMatchedContextIfMessageApplicableToMultiple() {
+        // Given
+        HttpMessage msg = createMessage();
+        Context matchCtxOne = mock(Context.class);
+        TechSet expectedTechSet = new TechSet(Tech.Db);
+        given(matchCtxOne.getTechSet()).willReturn(expectedTechSet);
+        Context matchCtxTwo = mock(Context.class);
+        given(session.getContextsForUrl(msg.getRequestHeader().getURI().toString()))
+                .willReturn(asList(matchCtxOne, matchCtxTwo));
+        // When
+        PassiveScanData psd = new PassiveScanData(msg);
+        // Then
+        assertThat(psd.getTechSet(), is(equalTo(expectedTechSet)));
```

Comment:
I tried to do a similar test for getUsers but couldn't get it sorted out. psd can be mocked since it's final, and trying to do given//willReturn on an instantiated one will returned an empty list despite trying to give it a list to return.

Reply:
Add a class variable for the extension and allow to set it, it can then be mocked to return what's needed.